### DISCUSSION TITLE
Spy some dependencies

### DISCUSSION THREAD

**Discussion by jim-king-2000** [google/googletest#4510] [Community Help]
Suppose we have a function in my production code.

```C++
void ProductionCode()
{
    do1Thing();
    do2Thing();
    do3Thing();
}
```

Now we need write a unit test for this function. And we want to spy `do2Thing()` and let it throw an exception. Pseudo-code is like this:

```C++
#include "ProductionCode.h"

TEST()
{
    mock("do2Thing", [](){ throw "exception"; });
    ProductionCode();
    mock.assertThrow("exception");
}
```

Is it possible to do this?

**Comment by derekmauro** [comment]
Please see https://google.github.io/googletest/gmock_cook_book.html#mocking-free-functions.

**Reply by jim-king-2000** [reply]
Thanks for the quick reply. It seems that there is no real spy functionality in c++ as it is in other programming languages. We have to change our programming model for doing unit test.

**Comment by jim-king-2000** [comment]
Here is the pseudo-code which is similar to our production code in practice:
```C++
char *do1Thing()
{
    char *info = new char[64];
    // populate info with some important information
    return info;
}

void do2Things()
{
    // do some dangerous operations here
}

void do3Thing(char *info)
{
    // use info here
    delete info;
}

void ProductionCode()
{
    char *info = do1Thing();
    do2Thing();
    do3Thing(info);
}
```
In the code above, we create a dynamic memory in `do1Thing()` and we release it in `do3Thing()`.

Now we need to write a unit test. The case is to "make sure all of the resources are released when any exception is thrown". So we need invoke `ProductionCode()` in our test code and let `do2Thing()` throw an exception. Obviously, in the production code, `do2Thing()` seldom fails. That is why we need mock it here.

```C++
#include "ProductionCode.h"

TEST()
{
    mock("do2Thing", [](){ throw "exception"; });
    ProductionCode();
    mock.assertThrow("exception");
}
```
Here we write a fake `do2Thing` and let it throw an exception. Then we call `ProductionCode()` and expect it could invoke the mocked version of `do2Thing()`. If it were true, we can see memory leak when running this case with valgrind or other memory debugger.

In other languages, such as node.js, we can do it like this. Every function including the system function could be mocked. Does this way work in C++? If no, what is the best practice?

See the mock of node.js native test:
```Javascript
import { mock } from "node:test";

const tt = {
  f0() {
    console.log("f0");
    this.f1();
  },

  f1() {
    console.log("f1");
  },
};

mock.method(tt, "f1", () => console.log("mock"));
tt.f0();
```

The output is:
```console
f0
mock
```
It means that `tt.f1()` has been successfully mocked. Is there any mechanism similar to the one of node.js?
